### ISSUE TITLE
cmd/ebitenmobile: multitouch releases all touches at once on Android

### ISSUE THREAD

**Issue by divVerent** [hajimehoshi/ebiten#2220]
NOTE: this happens ONLY when using an Android Studio project and "ebitenmobile bind"; when using the "gomobile install" method this issue does NOT happen!

When I run the "paint" example via ebitenmobile, touches behave just fine on my Pixel 6; pressing one, then another finger, then releasing the second finger does not release the first one. However, in my game precisely this happens.

I've traced this to the root cause:

- When using gomobile install, touches are received by `internal/ui/ui_mobile.go` in form of `TypeBegin`/... events. This part works fine.
- When using ebitenmobile bind, touches are received by a Java function in cmd/ebitenmobile/gobind.go:

```
    @Override
    public boolean onTouchEvent(MotionEvent e) {
        for (int i = 0; i < e.getPointerCount(); i++) {
            int id = e.getPointerId(i);
            int x = (int)e.getX(i);
            int y = (int)e.getY(i);
            Ebitenmobileview.updateTouchesOnAndroid(e.getActionMasked(), id, (in
t)pxToDp(x), (int)pxToDp(y));
        }
        return true;
    }
```

The problem is that this does not seem to be how touch events work - on my device, if I press two fingers then release one, I get a MotionEvent of type ACTION_POINTER_UP that contains BOTH fingers! Thus my game thinks both fingers were released.

Going by Google's documentation, the fix is only calling updateTouchesOnAndroid for all pointers if the event is a ACTION_MOVE event; batching is only the case for those according to https://developer.android.com/reference/android/view/MotionEvent, and otherwise one apparently should call getActionIndex() and only move the pointer up/down that is explicitly indicated by that index (i.e. getPointerId(getActionIndex()).

It seems like gomobile considers the additional fingers still moves - see here:

https://cs.opensource.google/go/x/mobile/+/master:app/android.go;l=419;drc=e579adbbc4a2bb7b98a904201d0200df2f08c2be

I can't tell from the docs if this is right or not, but it appears to be harmless, so I'd suggest Ebitengine should just do the same and remap all non-getActionIndex() fingers to ACTION_MOVE.

I'd send a patch for this myself, but can't seem to figure out how to get the changes into gobind.src.go, so for now I'm just proposing an untested patch here:

```
diff --git a/cmd/ebitenmobile/gobind.go b/cmd/ebitenmobile/gobind.go
index 3760f99e..5e63f59d 100644
--- a/cmd/ebitenmobile/gobind.go
+++ b/cmd/ebitenmobile/gobind.go
@@ -419,11 +419,16 @@ public class EbitenView extends ViewGroup implements InputManager.InputDeviceLis
 
     @Override
     public boolean onTouchEvent(MotionEvent e) {
+        // If a finger went up or down, this is its index.
+        // All others may have moved, or not.
+        // Android does not batch multiple fingers going up/down together.
+        int touchIndex = e.getActionIndex();
         for (int i = 0; i < e.getPointerCount(); i++) {
             int id = e.getPointerId(i);
             int x = (int)e.getX(i);
             int y = (int)e.getY(i);
-            Ebitenmobileview.updateTouchesOnAndroid(e.getActionMasked(), id, (int)pxToDp(x), (int)pxToDp(y));
+            int action = (i == touchIndex) ? e.getActionMasked() : MotionEvent.ACTION_MOVE;
+            Ebitenmobileview.updateTouchesOnAndroid(action, id, (int)pxToDp(x), (int)pxToDp(y));
         }
         return true;
     }
```

**Comment by divVerent** [comment]
I just noticed that my patch is not quite correct - this would remap some extra actions that we're normally ignoring to ACTION_MOVE. This is probably harmless for fullscreen games, but maybe not quite. Might be better to do the filtering for the action types we care about in Java then?

**Comment by divVerent** [comment]
I just noticed that my patch is not quite correct - this would remap some extra actions that we're normally ignoring to ACTION_MOVE. This is probably harmless for fullscreen games, but maybe not quite. Might be better to do the filtering for the action types we care about in Java then?

**Comment by hajimehoshi** [comment]
Thanks, I'll test your suggestion later soon.

**Comment by hajimehoshi** [comment]
> I'd send a patch for this myself, but can't seem to figure out how to get the changes into gobind.src.go, so for now I'm just proposing an untested patch here:

Updating `gobind.go` and then running `go generate ./cmd/ebitenmobile` should work

**Comment by hajimehoshi** [comment]
If you suceed updating gobind.go and create a PR, I'd be happy to review.

**Comment by hajimehoshi** [comment]
I'll try to fix this

**Comment by hajimehoshi** [comment]
https://developer.android.com/reference/android/view/MotionEvent#getActionMasked()

> Return the masked action being performed, without pointer index information. Use [getActionIndex()](https://developer.android.com/reference/android/view/MotionEvent#getActionIndex()) to return the index associated with pointer actions.

I think this is it.

**Comment by divVerent** [comment]
Yes, the patch seems correct enough and treating all other events as moves is what gomobile does too, so it should be fine - thanks! Soon trying on my device.

(BTW, my main issue was how to make it so I can run this file2byteslice tool - as your code is currently, I'd have to install that system-wide which I don't quite want... looking into that)

**Comment by divVerent** [comment]
What was the problem with the fix? The message merely says "limited impact" - anything missing in the fix?

**Comment by hajimehoshi** [comment]
I mean I don't back-port this fix to the stable branch as I want to apply only serious bug fixes there.

**Comment by hajimehoshi** [comment]
@corfe83 reported a suspicious issue with multitaps on an Android device. If this is the same issue, I'll backport the fix to v2.3 again.

**Comment by hajimehoshi** [comment]
As #2252 will be included in v2.3, let's add the fix for this issue to v2.3 too.
